### PR TITLE
Matmul BFloat16 support

### PR DESCRIPTION
**Pull request by zasdfgbnm**

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build

File: csrc/ops/arith.cpp [hunk 1]
```diff
@@ -2358,8 +2358,12 @@ TensorView* fusedMultiplySum(
   // TODO:
   //  Add tf32 and other mma data types
   //  Add fallback path for non-mma data types.
-  TORCH_CHECK(tv_a->getDataType().value() == DataType::Half);
-  TORCH_CHECK(tv_b->getDataType().value() == DataType::Half);
+  TORCH_CHECK(
```
Comment (drzejan2) [inline csrc/ops/arith.cpp:2361]: Based on the added test, the goal is to support for problems where bf16 is the type of both inputs. The checks here will let the following mixed precision (`tv_a` -> bf16, `tv_b` -> fp16).

How about adding one more check:
```
TORCH_CHECK(tv_b->getDataType().value() == tv_b->getDataType().value() );
```
?

**Comment by zasdfgbnm** [comment]
!build

File: csrc/ops/arith.cpp (hunk 1 above)
Comment (zasdfgbnm) [reply csrc/ops/arith.cpp:2361]: changed

**Review by naoyam — APPROVED** [review]
Seems mostly mechanical. LGTM

**Comment by rdspring1** [comment]
I'm seeing some compile errors in `benchmark/matmul.cpp`:
```cpp
FAILED: CMakeFiles/nvfuser_bench.dir/benchmark/matmul.cpp.o
In function ‘void SingleMatmulBase(benchmark::State&, MatmulLayout, nvfuser::MatmulParams)’:
/home/rspring/workspace/top_level_nvfuser/benchmark/matmul.cpp:132:17: error: ‘fp16MatmulAtInput’ was not declared in this scope
  132 |   auto inputs = fp16MatmulAtInput(
      |                 ^~~~~~~~~~~~~~~~~
/home/rspring/workspace/top_level_nvfuser/benchmark/matmul.cpp:137:79: error: cannot convert ‘<brace-enclosed initializer list>’ to ‘const c10::ArrayRef<c10::IValue>&’
  137 |   KernelArgumentHolder args = KernelArgumentHolder::createKernelArgumentHolder(
      |                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
  138 |       {inputs.first, inputs.second});
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/rspring/workspace/top_level_nvfuser/csrc/executor_utils.h:19,
                 from /home/rspring/workspace/top_level_nvfuser/csrc/executor.h:10,
                 from /home/rspring/workspace/top_level_nvfuser/benchmark/matmul.cpp:8:
/home/rspring/workspace/top_level_nvfuser/csrc/executor_kernel_arg.h:332:41: note:   initializing argument 1 of ‘static nvfuser::KernelArgumentHolder nvfuser::KernelArgumentHolder::createKernelArgumentHolder(const c10::ArrayRef<c10::IValue>&)’
  332 |       const c10::ArrayRef<c10::IValue>& inputs);
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~
/home/rspring/workspace/top_level_nvfuser/benchmark/matmul.cpp:159:30: error: no matching function for call to ‘nvfuser::FusionExecutor::runFusion(<brace-enclosed initializer list>)’
  159 |   auto outputs = fe.runFusion({inputs.first, inputs.second});
      |                  ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /home/rspring/workspace/top_level_nvfuser/benchmark/matmul.cpp:8:
/home/rspring/workspace/top_level_nvfuser/csrc/executor.h:99:27: note: candidate: ‘std::vector<at::Tensor> nvfuser::FusionExecutor::runFusion(nvfuser::KernelArgumentHolder&, const nvfuser::LaunchParams&, nvfuser::CompileParams, std::vector<at::Tensor>)’
   99 |   std::vector<at::Tensor> runFusion(
      |                           ^~~~~~~~~
/home/rspring/workspace/top_level_nvfuser/csrc/executor.h:100:29: note:   no known conversion for argument 1 from ‘<brace-enclosed initializer list>’ to ‘nvfuser::KernelArgumentHolder&’
  100 |       KernelArgumentHolder& args,
      |       ~~~~~~~~~~~~~~~~~~~~~~^~~~
/home/rspring/workspace/top_level_nvfuser/csrc/executor.h:105:27: note: candidate: ‘std::vector<at::Tensor> nvfuser::FusionExecutor::runFusion(const c10::ArrayRef<c10::IValue>&, const std::vector<at::Tensor>&, const nvfuser::LaunchParams&, nvfuser::CompileParams, const c10::optional<long unsigned int>&)’
  105 |   std::vector<at::Tensor> runFusion(
      |                           ^~~~~~~~~
/home/rspring/workspace/top_level_nvfuser/csrc/executor.h:105:27: note:   candidate expects 5 arguments, 1 provided
/home/rspring/workspace/top_level_nvfuser/csrc/executor.h:119:27: note: candidate: ‘std::vector<at::Tensor> nvfuser::FusionExecutor::runFusion(const c10::ArrayRef<c10::IValue>&, const nvfuser::LaunchParams&, nvfuser::CompileParams, const c10::optional<long unsigned int>&)’
  119 |   std::vector<at::Tensor> runFusion(
      |                           ^~~~~~~~~
/home/rspring/workspace/top_level_nvfuser/csrc/executor.h:120:40: note:   no known conversion for argument 1 from ‘<brace-enclosed initializer list>’ to ‘const c10::ArrayRef<c10::IValue>&’
  120 |       const at::ArrayRef<c10::IValue>& inputs,
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~
/home/rspring/workspace/top_level_nvfuser/benchmark/matmul.cpp:166:32: error: no matching function for call to ‘nvfuser::FusionExecutor::runFusion(<brace-enclosed initializer list>)’
  166 |     auto outputs = fe.runFusion({inputs.first, inputs.second});
      |                    ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by zasdfgbnm** [comment]
https://github.com/NVIDIA/Fuser/pull/239

**Pull request merged**